### ISSUE TITLE
VLAN 0 on some Portchannel

### ISSUE THREAD

**Issue by matrix1000** [netdisco/netdisco#581]
Hi Netdisco founds some VLAN 0 on some Portchannels?

## Expected Behavior
there shouldn't be any VLAN 0
:-)

## Current Behavior
I see vlan 0 on 4 different cisco 38xx Stacks.
But on every stack only once.
Netdisco see the vlan 0 as native vlan on the Trunked Portchannel.

## Possible Solution
sorry, no idea

## Steps to Reproduce (for bugs)
1. 
2. 
3. 
4. 

## Context
This issue does not affected much.

## Your Environment
* Netdisco version used: 2.42.8
* SNMP::Info version used: 3.68
122 Devices 
and only on 4 Devices I have this strange problem.

Many Thanks for Help

**Comment by matrix1000** [comment]
Hi,

Can I help with any additional information, so solve the problem?

**Comment by rc9000** [comment]
I can confirm this issue, at a quick glance we have around 70 VLAN  0, originating mostly from these platforms:

   3560E24TD
   3560G24TS
   37xxStack
   38xxstack

Macsuck logs something like 

    [30830] 2019-05-24 10:29:02 debug  [172.28.167.151] macsuck - VLANs: 0,1,100,1002,1003,1004,1600,1601
    [30830] 2019-05-24 10:29:02 debug  [172.28.167.151] macsuck - invalid VLAN number 0
    [30830] 2019-05-24 10:29:05 debug  [172.28.167.151] macsuck - port Port-channel1 vlan 0 : 1 nodes

My first guess would be buggy SNMP replies from Cisco. Unfortunately I can't dig deeply into this right now, but maybe this helps somebody else or might get around to it later.

**Comment by inphobia** [comment]
i see this too on c3750x, stack or no stack, port channel or single port don't seem to matter (using SNMP::Info::Layer3::C6500), snmpv2.

however i can't reproduce your vlan listing, mine does not list vlan 0:
```
[17271] 2019-05-26 00:37:31 debug  [10.40.254.12] macsuck - VLANs: 1,1002,1003,1004,1005,102,105,110,111,112,113,115,120,252,254,451,452,455,460,8,812,830
[17271] 2019-05-26 00:37:31 debug  [10.40.254.12] macsuck VLAN 1002 - skipped by macsuck_no_vlan config
[17271] 2019-05-26 00:37:31 debug  [10.40.254.12] macsuck VLAN 1003 - skipped by macsuck_no_vlan config
[17271] 2019-05-26 00:37:31 debug  [10.40.254.12] macsuck VLAN 1004 - skipped by macsuck_no_vlan config
[17271] 2019-05-26 00:37:31 debug  [10.40.254.12] macsuck VLAN 1005 - skipped by macsuck_no_vlan config
```

it goes over all vlans with snmp reindexing (short extract:)
```
[17271] 2019-05-26 00:37:47 debug [10.40.254.12] reindexing to vlan 111 (ver: 2, class: SNMP::Info::Layer3::C6500)
[17271] 2019-05-26 00:37:47 debug  [10.40.254.12] macsuck 6c:b2:ae:91:25:13 - port Port-channel1 has neighbor 10.1.100.251 - skipping.
[17271] 2019-05-26 00:37:47 debug  [10.40.254.12] macsuck 00:26:f0:0b:00:00 - port Port-channel1 has neighbor 10.1.100.251 - skipping.
[17271] 2019-05-26 00:37:47 debug  [10.40.254.12] check_mac - HSRP mac [00:00:0c:9f:f0:6f] - skipping
[17271] 2019-05-26 00:37:47 debug [10.40.254.12] reindexing to vlan 112 (ver: 2, class: SNMP::Info::Layer3::C6500)
[17271] 2019-05-26 00:37:48 debug  [10.40.254.12] macsuck 00:e0:67:0f:4c:8f - port Port-channel1 has neighbor 10.1.100.251 - skipping.
[17271] 2019-05-26 00:37:48 debug  [10.40.254.12] macsuck 00:20:6b:e9:df:e8 - port Port-channel1 has neighbor 10.1.100.251 - skipping.
[17271] 2019-05-26 00:37:48 debug  [10.40.254.12] macsuck 00:e0:67:0f:44:05 - port Port-channel1 has neighbor 10.1.100.251 - skipping.
```

at the end it reindexes to vlan 0 for some reason:
```
[17271] 2019-05-26 00:38:13 debug [10.40.254.12] reindexing to vlan 0 (ver: 2, class: SNMP::Info::Layer3::C6500)
[17271] 2019-05-26 00:38:13 debug  [10.40.254.12] macsuck - port GigabitEthernet4/0/21 vlan 812 : 1 nodes
[17271] 2019-05-26 00:38:13 debug  [10.40.254.12] macsuck - port GigabitEthernet2/0/23 vlan 812 : 1 nodes
[17271] 2019-05-26 00:38:13 debug  [10.40.254.12] macsuck - port GigabitEthernet1/0/29 vlan 812 : 1 nodes
...
[17271] 2019-05-26 00:38:14 debug  [10.40.254.12] macsuck - port GigabitEthernet2/0/5 vlan 1 : 3 nodes
[17271] 2019-05-26 00:38:14 debug  [10.40.254.12] macsuck - port GigabitEthernet2/0/5 vlan 0 : 0 nodes
[17271] 2019-05-26 00:38:14 debug  [10.40.254.12] macsuck - port GigabitEthernet3/0/8 vlan 0 : 0 nodes
[17271] 2019-05-26 00:38:14 debug  [10.40.254.12] macsuck - 298 updated forwarding table entries
[17271] 2019-05-26 00:38:14 debug  [10.40.254.12] macsuck - removed 289 fwd table entries to archive
```
and eventually finds 2 ports with vlan 0 but also with 0 nodes.

try as i want i can't find a vlan 0 with any of the functions.
```
            i_up_admin                                 "ifAdminStatus",
            i_vlan_stat                                "vmPortStatus",
            i_vlan_type                                "vmVlanType",
            i_vlan_1                                   "vmVlans",
            i_vlan_2                                   "vmVlans2k",
            i_vlan_3                                   "vmVlans3k",
            i_vlan_4                                   "vmVlans4k",
            i_vlan2                                    "vmVlan",
            i_voice_vlan                               "vmVoiceVlanId",


SNMP::Info::_load_attr i_index : IF-MIB::ifIndex : .1.3.6.1.2.1.2.2.1.1
SNMP::Info::_load_attr i_description : IF-MIB::ifDescr : .1.3.6.1.2.1.2.2.1.2
SNMP::Info::_load_attr qb_fw_port : Q-BRIDGE-MIB::dot1qTpFdbPort : .1.3.6.1.2.1.17.7.1.2.2.1.2
SNMP::Info::_load_attr orig_fw_mac : BRIDGE-MIB::dot1dTpFdbAddress : .1.3.6.1.2.1.17.4.3.1.1
SNMP::Info::_load_attr qb_fw_port : Q-BRIDGE-MIB::dot1qTpFdbPort : .1.3.6.1.2.1.17.7.1.2.2.1.2
SNMP::Info::_load_attr orig_fw_port : BRIDGE-MIB::dot1dTpFdbPort : .1.3.6.1.2.1.17.4.3.1.2
SNMP::Info::_load_attr qb_fw_port : Q-BRIDGE-MIB::dot1qTpFdbPort : .1.3.6.1.2.1.17.7.1.2.2.1.2
SNMP::Info::_load_attr dot1qVlanFdbId : Q-BRIDGE-MIB::dot1qVlanFdbId : .1.3.6.1.2.1.17.7.1.4.2.1.3
SNMP::Info::_load_attr bp_index : BRIDGE-MIB::dot1dBasePortIfIndex : .1.3.6.1.2.1.17.1.4.1.2
SNMP::Info::_load_attr i_type : IF-MIB::ifType : .1.3.6.1.2.1.2.2.1.3
SNMP::Info::_load_attr vtp_trunk_native : CISCO-VTP-MIB::vlanTrunkPortNativeVlan : .1.3.6.1.4.1.9.9.46.1.6.1.1.5
SNMP::Info::_load_attr i_vlan2 : CISCO-VLAN-MEMBERSHIP-MIB::vmVlan : .1.3.6.1.4.1.9.9.68.1.2.2.1.2
SNMP::Info::_load_attr vtp_trunk_dyn : CISCO-VTP-MIB::vlanTrunkPortDynamicState : .1.3.6.1.4.1.9.9.46.1.6.1.1.13
SNMP::Info::_load_attr vtp_trunk_dyn_stat : CISCO-VTP-MIB::vlanTrunkPortDynamicStatus : .1.3.6.1.4.1.9.9.46.1.6.1.1.14
SNMP::Info::_load_attr v_cvi_if : CISCO-VLAN-IFTABLE-RELATIONSHIP-MIB::cviRoutedVlanIfIndex : .1.3.6.1.4.1.9.9.128.1.1.1.1.3
SNMP::Info::_load_attr vtp_trunk_vlans : CISCO-VTP-MIB::vlanTrunkPortVlansEnabled : .1.3.6.1.4.1.9.9.46.1.6.1.1.4
SNMP::Info::_load_attr vtp_trunk_vlans_2k : CISCO-VTP-MIB::vlanTrunkPortVlansEnabled2k : .1.3.6.1.4.1.9.9.46.1.6.1.1.17
SNMP::Info::_load_attr vtp_trunk_vlans_3k : CISCO-VTP-MIB::vlanTrunkPortVlansEnabled3k : .1.3.6.1.4.1.9.9.46.1.6.1.1.18
SNMP::Info::_load_attr vtp_trunk_vlans_4k : CISCO-VTP-MIB::vlanTrunkPortVlansEnabled4k : .1.3.6.1.4.1.9.9.46.1.6.1.1.19
SNMP::Info::_load_attr v_state : CISCO-VTP-MIB::vtpVlanState : .1.3.6.1.4.1.9.9.46.1.3.1.1.2
SNMP::Info::_load_attr i_voice_vlan : CISCO-VLAN-MEMBERSHIP-MIB::vmVoiceVlanId : .1.3.6.1.4.1.9.9.68.1.5.1.1.1
SNMP::Info::_load_attr v_name : CISCO-VTP-MIB::vtpVlanName : .1.3.6.1.4.1.9.9.46.1.3.1.1.4
```

interface config is bog standard:
```
interface GigabitEthernet3/0/8
 switchport mode access
 switchport voice vlan 812
 srr-queue bandwidth share 1 30 35 5
 priority-queue out
 mls qos trust cos
 auto qos trust
 storm-control broadcast level 20.00
 storm-control multicast level 20.00
 spanning-tree portfast
end
```

i do have the following in my config:
```
macsuck_no_vlan:
  - 'fddi-default'
  - 'token-ring-default'
  - 'fddinet-default'
  - 'trnet-default'
  - 'fcoe-vsan-4048'
  - 'SAM-vlan-boot'
  - 'SAM-vlan-appliance-management'
  - 'SAM-vlan-management'

devport_vlan_limit: 250
```

wonder where you guys are getting vlan 0 from in your vlan listing.



for the nodes on vlan 0, that's a feature it seems:
https://github.com/netdisco/netdisco/blob/065e74df62db74064776b0b40fd65f31a035e1d1/lib/App/Netdisco/Worker/Plugin/Macsuck/Nodes.pm#L54

https://github.com/netdisco/netdisco/blob/065e74df62db74064776b0b40fd65f31a035e1d1/lib/App/Netdisco/Worker/Plugin/Macsuck/Nodes.pm#L103-L104


anything related to vlans in your config?

**Comment by ollyg** [comment]
Hold on, hold on. Vlan 0 is not an error from the device. It's used as a placeholder internally in Netdisco. There are good reasons for that, so arguably this is a user interface issue as the vlan 0 should not show in the UI.

The reason for using vlan 0 comes from the code in Worker::Plugin::Macsuck::Nodes, when the device reports a known MAC but does not say which port the MAC is active on. Why do this? Because some devices go the other way and report the MAC twice, especially when using per-vlan SNMP indexing. It's just cleaner for the code, and for the database, to remain numeric all the time.

So in summary vlan 0 can be in the DB but should be treated as null/none for the web interface.

**Comment by ollyg** [comment]
@matrix1000 please can you post some screenshots of the web UI showing vlan 0? Thanks.

**Comment by matrix1000** [comment]
Hi,

Many thanks.
yes, here are some screenshots. Hopefully they help.:
![vlan0-1](https://user-images.githubusercontent.com/1917331/58454397-a04e5180-811e-11e9-9168-a67771f0956f.png)
![vlan0-2](https://user-images.githubusercontent.com/1917331/58454398-a04e5180-811e-11e9-9407-f575f8830084.png)
![vlan0-3](https://user-images.githubusercontent.com/1917331/58454399-a04e5180-811e-11e9-8890-9501e4f2e30a.png)

**Comment by inphobia** [comment]
strange, why is your port channel reported as down?

**Comment by matrix1000** [comment]
Because it is a portchannel without members, therefor it is down.

```
sh etherchannel summary 
Number of channel-groups in use: 1
Number of aggregators:           1
Group  Port-channel  Protocol    Ports
------+-------------+-----------+-----------------------------------------------
1      Po1(SD)          -        
```
I'm starting to clean up my inherited company network. :-)

**Comment by inphobia** [comment]
are you running vtp by any chance?

**Comment by matrix1000** [comment]
no, vtp is disabled

**Comment by ollyg** [comment]
Thanks! this is fixed in:
[master 2c138a56] #581 skip storing and displaying fabricated vlan 0
